Reland "Add torch.utils.device_mode"

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91892
* __->__ #91796

Original PR https://github.com/pytorch/pytorch/pull/91525

Signed-off-by: Edward Z. Yang <ezyangfb.com>

ezyang committed:
Reland "Add torch.utils.device_mode"

Original PR https://github.com/pytorch/pytorch/pull/91525

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42381089).

albanD reviewed (commented):
Are you planning internal-only changes for this?

ezyang commented:
No, let me mark the difference

ezyang commented on torch/utils/_contextlib.py:
```diff
@@ -0,0 +1,143 @@
+                          "future versions. You should only decorate functions or methods. "
+                          "To preserve the current behavior of class decoration, you can "
+                          "directly decorate the `__init__` method and nothing else.")
+            func = cast(F, lambda *args, **kwargs: orig_func(*args, **kwargs))
```
Previously I overwrote func in the frame, now the closed over func has its own name

albanD commented on torch/_C/__init__.pyi.in:
```diff
@@ -45,6 +45,12 @@ class device:
     @overload
     def __init__(self, type: str, index: _int) -> None: ...
 
+    def __call__(self, func: T) -> T: ...
```
there is no call anymore.

ezyang committed:
Update on "Reland "Add torch.utils.device_mode""


Original PR https://github.com/pytorch/pytorch/pull/91525

Signed-off-by: Edward Z. Yang <ezyangfb.com>

Differential Revision: [D42381089](https://our.internmc.facebook.com/intern/diff/D42381089)

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented:
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42381089).

ezyang committed:
Update on "Reland "Add torch.utils.device_mode""


Original PR https://github.com/pytorch/pytorch/pull/91525

Signed-off-by: Edward Z. Yang <ezyangfb.com>

Differential Revision: [D42381089](https://our.internmc.facebook.com/intern/diff/D42381089)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42381089).

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge

ezyang commented on test/test_autograd.py:
```diff
@@ -122,22 +122,36 @@ def test_grad_mode_class_decoration(self):
+
+            # Show that we can actually construct the class
+            foo = Foo()
+            foo.foo()
```
Improved test here now catches the infinite loop bug from previous

huydhn commented:
@pytorchbot revert -m 'This breaks trunk with the following failed test https://hud.pytorch.org/failure/test_jit_save%2CTestTracer' -c landrace

ezyang committed:
Update on "Reland "Add torch.utils.device_mode""


Original PR https://github.com/pytorch/pytorch/pull/91525

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
These tests don't fail locally for me with the PR.

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge -f "all known trunk failures"
